### PR TITLE
Add replay functionality with tensor conversion fixes

### PR DESCRIPTION
**Pull request by LeonLiu4**
Trajectory recording during teleop (press 'r' to start/stop)
Standalone replay script with viewer support
Sample trajectory file for testing
Pick and place task with target visualization

### CODE REVIEW

File: examples/replay_with_viewer.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import sys
+import os
+import genesis as gs
+
+# Add the src directory to the path so we can import our modules
+sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'src'))
+
+from agent.gs_agent.wrappers.teleop_wrapper import TeleopWrapper
+from env.gs_env.sim.envs.so101_cube_env import SO101CubeEnv
+
```
Comment (yun-long) [inline examples/replay_with_viewer.py:15]: It is not the correct way of importing the python packages. It is recommended to use `uv` to first install `gs_env` via `uv sync --package gs-env`. Afterwards, you should be able to import `gs_agent` or `gs_env` directly without `sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'src'))`

```python
from gs_agent.wrappers.teleop_wrapper import TeleopWrapper
from gs_env.sim.envs.so101_cube_env import SO101CubeEnv
```

File: src/agent/gs_agent/wrappers/teleop_wrapper.py [hunk 2]
```diff
@@ -318,4 +350,175 @@ def close(self) -> None:
...
+
+        self.trajectory_data.append(step_data)
+
+    def _save_trajectory(self) -> None:
+        """Save trajectory data to disk."""
+        if not self.trajectory_data:
+            return
+
+        # Create trajectories directory if it doesn't exist
+        os.makedirs("trajectories", exist_ok=True)
```
Comment (yun-long) [inline src/agent/gs_agent/wrappers/teleop_wrapper.py:412]: avoid using ad-hoc text strings, it is error prone. Use constant instead. e.g., `TRAJECTOR_PATH = ...`

File: src/agent/gs_agent/wrappers/teleop_wrapper.py [hunk 3]
```diff
@@ -318,4 +350,175 @@ def close(self) -> None:
...
+
+        # Generate filename with timestamp
+        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
+        filename = f"trajectories/so101_pick_place_{timestamp}.json"
```
Comment (yun-long) [inline src/agent/gs_agent/wrappers/teleop_wrapper.py:416]: same here

File: src/agent/gs_agent/wrappers/teleop_wrapper.py [hunk 4]
```diff
@@ -318,4 +350,175 @@ def close(self) -> None:
...
+        if not os.path.exists(traj_file_path):
+            print(f"❌ Trajectory file not found: {traj_file_path}")
+            return
+
+        if not hasattr(self, '_teleop_env') or self._teleop_env is None:
+            print("❌ No environment set. Call set_environment() first.")
+            return
+
+        try:
+            # Load trajectory
```
Comment (yun-long) [inline src/agent/gs_agent/wrappers/teleop_wrapper.py:463]: same here

File: src/env/gs_env/sim/envs/so101_cube_env.py [hunk 5]
```diff
@@ -151,7 +151,12 @@ def _update_target_visualization(self) -> None:
         """Update target visualization to match robot end-effector."""
         try:
             pos, quat = self.entities["robot"].get_ee_pose()
-            if pos is not None:
+            if pos is not None and quat is not None:
```
Comment (yun-long) [inline src/env/gs_env/sim/envs/so101_cube_env.py:154]: same here. not sure what is the motivation of using `try-except` ?

File: src/env/gs_env/sim/robots/so101_robot.py [hunk 6]
```diff
@@ -204,8 +204,20 @@ def reset_to_pose(self, joint_angles: NDArray[np.float64]) -> None:
     def get_ee_pose(self) -> tuple[NDArray[np.float64] | None, NDArray[np.float64] | None]:
         """Get current end-effector pose."""
         try:
-            pos = np.array(self.ee_link.get_pos())
-            quat = np.array(self.ee_link.get_quat())
+            pos_tensor = self.ee_link.get_pos()
```
Comment (yun-long) [inline src/env/gs_env/sim/robots/so101_robot.py:207]: same here.

File: src/agent/gs_agent/wrappers/teleop_wrapper.py [hunk 7]
```diff
@@ -150,6 +165,10 @@ def step(self, action: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor, torch.
         else:
             obs = {}
 
+        # Record trajectory data if recording
+        if self.recording and command is not None:
+            self._record_trajectory_step(command, obs)
+
         # Return standard step format (empty tensors for compatibility)
         return (
             torch.tensor([]),          # next_obs
```
Comment (yun-long) [inline src/agent/gs_agent/wrappers/teleop_wrapper.py:174]: the `return` here doesn't have to be the same as the `gym-wrappe`r since `teleop_wrapper` doesn't require rewards. Reward signal and terminal signals are only required for reinforcement learning.

File: src/agent/gs_agent/wrappers/teleop_wrapper.py (hunk 2 above)
Comment (LeonLiu4) [reply src/agent/gs_agent/wrappers/teleop_wrapper.py]: fixed

File: src/agent/gs_agent/wrappers/teleop_wrapper.py [hunk 8]
```diff
@@ -318,4 +350,175 @@ def close(self) -> None:
...
+                "recording_date": datetime.now().isoformat(),
+                "duration_seconds": self.trajectory_data[-1]["timestamp"] if self.trajectory_data else 0,
+                "num_steps": len(self.trajectory_data),
+                "movement_speed": self.movement_speed,
+                "rotation_speed": self.rotation_speed
+            },
+            "trajectory": self.trajectory_data
+        }
+
+        try:
```
[the remark this answers is not in the document]
Comment (LeonLiu4) [reply src/agent/gs_agent/wrappers/teleop_wrapper.py]: removed

File: src/agent/gs_agent/wrappers/teleop_wrapper.py (hunk 4 above)
Comment (LeonLiu4) [reply src/agent/gs_agent/wrappers/teleop_wrapper.py]: removed

File: src/env/gs_env/sim/robots/so101_robot.py (hunk 6 above)
Comment (LeonLiu4) [reply src/env/gs_env/sim/robots/so101_robot.py]: removed

File: src/env/gs_env/sim/robots/so101_robot.py [hunk 9]
```diff
@@ -214,7 +226,11 @@ def get_ee_pose(self) -> tuple[NDArray[np.float64] | None, NDArray[np.float64] |
     def get_joint_positions(self) -> NDArray[np.float64]:
         """Get current joint positions."""
         try:
-            return self.entity.get_qpos().clone()
+            qpos_tensor = self.entity.get_qpos()
```
[the remark this answers is not in the document]
Comment (LeonLiu4) [reply src/env/gs_env/sim/robots/so101_robot.py]: removed

File: src/agent/gs_agent/wrappers/teleop_wrapper.py (hunk 7 above)
Comment (LeonLiu4) [reply src/agent/gs_agent/wrappers/teleop_wrapper.py]: kept the tensors to have the same format as other wrappers but now they are empty

File: src/env/gs_env/sim/robots/so101_robot.py [hunk 10]
```diff
@@ -214,7 +226,11 @@ def get_ee_pose(self) -> tuple[NDArray[np.float64] | None, NDArray[np.float64] |
...
+            if isinstance(qpos_tensor, torch.Tensor):
+                return qpos_tensor.cpu().numpy()
+            else:
+                return np.array(qpos_tensor)
         except Exception as e:
             print(f"Failed to get joint positions: {e}")
             return np.zeros(6)
```
[the remark this answers is not in the document]
Comment (LeonLiu4) [reply src/env/gs_env/sim/robots/so101_robot.py]: fixed
